File: DSA/saumya/Day9_Q1.py
class MaxStack:
    def __init__(self):
        self.main_stack = []
        self.max_stack = []

    def push(self, x):
        self.main_stack.append(x)
        if not self.max_stack or x >= self.max_stack[-1]:
            self.max_stack.append(x)

    def pop(self):
        if not self.main_stack:
            raise Exception("Stack is empty")
        popped_element = self.main_stack.pop()
        if popped_element == self.max_stack[-1]:
            self.max_stack.pop()

    def get_max(self):
        if not self.max_stack:
            raise Exception("Stack is empty")
        return self.max_stack[-1]


# Test the MaxStack class
stack = MaxStack()
stack.push(1)
stack.push(2)
print(stack.get_max())  # Output: 2
stack.pop()
print(stack.get_max())  # Output: 1

File: DSA/Ishita-Nigam/Day9/Day9_Q1.java
import java.util.Stack;

public class Day9_Q1 {

    static class MaxStack {
        private Stack<Integer> mainStack;
        private Stack<Integer> maxStack;

        public MaxStack() {
            mainStack = new Stack<>();
            maxStack = new Stack<>();
        }

        public void push(int x) {
            mainStack.push(x);

            if (maxStack.isEmpty() || x >= maxStack.peek()) {
                maxStack.push(x);
            }
        }

        public void pop() {
            if (mainStack.isEmpty()) {
                throw new RuntimeException("Stack is empty");
            }
            int poppedElement = mainStack.pop();

            if (poppedElement == maxStack.peek()) {
                maxStack.pop();
            }
        }

        public int get_max() {
            if (maxStack.isEmpty()) {
                throw new RuntimeException("Stack is empty");
            }
            return maxStack.peek();
        }
    }

    public static void main(String[] args) {
        MaxStack stack = new MaxStack();
        stack.push(1);
        stack.push(2);
        System.out.println(stack.get_max());
        stack.pop();
        System.out.println(stack.get_max());
    }
}
